group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/assignor/StickyTaskAssignor.java

```diff
@@ -0,0 +1,427 @@
+            }
+            processIdToState.get(member.processId).addTask(member.memberId, task, true);
+            it.remove();
+            updateHelpers(member, task, true);
+
+        }
+    }
+
+    private void maybeUpdateTasksPerMember(final int activeTasksNo) {
+        if (activeTasksNo == tasksPerMember) {
+            totalCapacity--;
+            allTasks -= activeTasksNo;
+            tasksPerMember = computeTasksPerMember(allTasks, totalCapacity);
+        }
+    }
+
+    private Member findMemberWithLeastLoad(final Set<Member> members, TaskId taskId, final boolean returnSameMember) {
+        if (members == null || members.isEmpty()) {
+            return null;
+        }
+        Set<Member> rightPairs = members.stream()
+            .filter(member  -> taskPairs.hasNewPair(taskId, processIdToState.get(member.processId).assignedTasks()))
+            .collect(Collectors.toSet());
+        if (rightPairs.isEmpty()) {
+            rightPairs = members;
+        }
+        Optional<ProcessState> processWithLeastLoad = rightPairs.stream()
+            .map(member  -> processIdToState.get(member.processId))
+            .min(Comparator.comparingDouble(ProcessState::load));
+
+        // processWithLeastLoad must be present at this point, but we do a double check
+        if (processWithLeastLoad.isEmpty()) {
+            return null;
+        }
+        // if the same exact former member is needed
+        if (returnSameMember) {
+            return standbyTaskToPrevMember.get(taskId).stream()
+                .filter(standby -> standby.processId.equals(processWithLeastLoad.get().processId()))
+                .findFirst()
+                .orElseGet(() -> memberWithLeastLoad(processWithLeastLoad.get()));
+        }
+        return memberWithLeastLoad(processWithLeastLoad.get());
+    }
+
+    private Member memberWithLeastLoad(final ProcessState processWithLeastLoad) {
+        Optional<String> memberWithLeastLoad = processWithLeastLoad.memberToTaskCounts().entrySet().stream()
+            .min(Map.Entry.comparingByValue())
+            .map(Map.Entry::getKey);
+        return memberWithLeastLoad.map(memberId -> new Member(processWithLeastLoad.processId(), memberId)).orElse(null);
+    }
+
+    private boolean hasUnfulfilledQuota(final Member member) {
+        return processIdToState.get(member.processId).memberToTaskCounts().get(member.memberId) < tasksPerMember;
+    }
+
+    private void assignStandby(final Set<TaskId> standbyTasks, final int numStandbyReplicas) {
+        for (TaskId task : standbyTasks) {
+            for (int i = 0; i < numStandbyReplicas; i++) {
+
+                final Set<String> availableProcesses = processIdToState.values().stream()
+                    .filter(process -> !process.hasTask(task))
+                    .map(ProcessState::processId)
+                    .collect(Collectors.toSet());
+
+                final String errorMessage = "Unable to assign " + (numStandbyReplicas - i) +
```

Comment:
It would probably be good to move the error message construction to a private method, so that we do not have to do this string concatenation in every iteration of the loop even if no error occurs.